Unique email sign in without specifying tenant (#205)
aspnetzero/documents: docs/en/Core-Angular-Sign-In-Without-Specifying-Tenant.md

## Sign In Without Specifying Tenant

> Normally, **ASP.NET Zero** uses tenant information in login transactions. This document shows you how to implement tenant information independent login. 

Reviewer: ```suggestion
Normally, **ASP.NET Zero** uses tenant information during the login process. This document shows you how to implement the login process without tenant information. 
```

> ***Important Note:*** *To implement that, your user's email addresses have to be unique. Otherwise, that solution may not works properly.*

Reviewer: ```suggestion
***Important Note:*** *Your user's email addresses have to be unique to implement this solution. Otherwise, this solution may not work correctly.*
```

#### Updating LogInManager

* First of all, open `LogInManager`. *(It is located in **aspnet-core\src\\[YOURAPPNAME].Application\Authorization** folder.)*

> * Add lines of codes shown below

Reviewer: ```suggestion
* Add lines shown below
```

      {
          throw new ArgumentException("login");
      }
      using (UnitOfWorkManager.Current.DisableFilter(AbpDataFilters.MayHaveTenant))
      {
          var user = await _userStore.FindAsync(login);
          if (user == null)
          {
              return new AbpLoginResult<Tenant, User>(AbpLoginResultType.UnknownExternalLogin);
          }
          //Get and check tenant
          Tenant tenant = null;
          if (!MultiTenancyConfig.IsEnabled)
          {
              tenant = await GetDefaultTenantAsync();
          }
          else if (user.TenantId.HasValue)
          {
              tenant = await TenantRepository.FirstOrDefaultAsync(t => t.Id == user.TenantId);
              if (tenant == null)
              {
                  return new AbpLoginResult<Tenant, User>(AbpLoginResultType.InvalidTenancyName);
              }
              if (!tenant.IsActive)
              {
                  return new AbpLoginResult<Tenant, User>(AbpLoginResultType.TenantIsNotActive, tenant);
              }
          }
          return await CreateLoginResultAsync(user, tenant);
      }
  }
  ``````

>   Then, your `LogInManager` will be able to use given user's tenant to try log in.

Reviewer: ```suggestion
  Then, your `LogInManager` will be able to use given user's tenant for login process.
```

  
  
  
  #### Updating UserManager
  
* Go to `UserManager`. *(It is located in **aspnet-core\src\\[YOURAPPNAME].Core\Authorization\Users** folder.)*

> * And add following lines of codes

Reviewer: ```suggestion
* And add following lines;
```

                  );
              }
              
              var refreshToken = CreateRefreshToken(await CreateJwtClaims(loginResult.Identity,
                  loginResult.User, tokenType: TokenType.RefreshToken)
              );
              
              var accessToken = CreateAccessToken(await CreateJwtClaims(loginResult.Identity,
                  loginResult.User, refreshTokenKey: refreshToken.key));
              
              return new ExternalAuthenticateResultModel
              {
                  AccessToken = accessToken,
                  EncryptedAccessToken = GetEncryptedAccessToken(accessToken),
                  ExpireInSeconds = (int) _configuration.AccessTokenExpiration.TotalSeconds,
                  RefreshToken = refreshToken.token,
                  RefreshTokenExpireInSeconds = (int) _configuration.RefreshTokenExpiration.TotalSeconds
              };
          }
          default:
          {
              throw _abpLoginResultTypeHelper.CreateExceptionForFailedLoginAttempt(
                  loginResult.Result,
                  model.ProviderKey,
                  loginResult?.Tenant?.Name
              );
          }
      }
  }
  ```

> Then your project will be able to use without specifying tenant.

Reviewer: ```suggestion
Then your users will be able to login without specifying a tenant.
```

#### More

For a more stable UI, you can remove the tenant selection model used for login operations.

> Go to **aspnet-zero-core\angular\src\account\account.component.ts** and add add **login** to **tenantChangeDisabledRoutes**

Reviewer: ```suggestion
Go to **aspnet-zero-core\angular\src\account\account.component.ts** and add **login** to **tenantChangeDisabledRoutes**
```
